### PR TITLE
Halving/Doubling style array vector operations

### PR DESCRIPTION
**Pull request by daviditen**
Implement the array vector operations using a buffer that grows and shrinks
by a factor of its current size instead of one element at a time.  Currently,
the growth factor defaults to 1.5, but it can be changed with a config
param.

Updated timeVectorArray.chpl test to include testing the remove() function.

Update domainMaps.chpl memory tracking test .goods.  The array-as-vec work
added a new domain member to DefaultRectangular arrays.  This test tracks
leaked memory from operations on different domain maps, and in some cases
DefaultRectangular arrays are leaked in it.  Since the leaked arrays have
a new domain field, the amount leaked for each leaked array went up.

### CODE REVIEW

File: modules/internal/ChapelArray.chpl [hunk 1]
```diff
@@ -2027,7 +2030,28 @@ module ChapelArray {
...
     }
-
+    pragma "no doc"
+    proc resizeAllocDomain(d: domain, d2: domain, factor=arrayAsVecGrowthFactor, direction=1, grow=1) {
+      const lo = d.low,
+            hi = d.high,
+            size = hi - lo + 1;
+      var newSize: int;
+      if grow > 0 {
+        newSize = max(size+1, (size*factor):int); // Always grow by at least 1.
```
Comment (ronawho) [inline modules/internal/ChapelArray.chpl]: Doesn't have to be done for this commit, but I think this should use `chpl_mem_good_alloc_size()` to reduce slack. All this really means is that the size will be rounded up to the actual size the allocator will use

https://github.com/facebook/folly/blob/master/folly/docs/FBVector.md#the-jemalloc-connection is a good section on why using `chpl_mem_good_alloc_size()` makes sense and that doc is really good general info on implementing a fast vector.

Comment (daviditen) [inline modules/internal/ChapelArray.chpl]: I read the FBVector article and used the 1.5 growth factor out of it.  I planned to leave the allocator specifics for another commit.  chpl_mem_good_alloc_size() is new enough that I hadn't thought to use it here though yet, so thanks for the tip.

Comment (ronawho) [inline modules/internal/ChapelArray.chpl]: In a separate commit makes sense to me. And sorry `chpl_mem_good_alloc_size` is the runtime routine. The chapel wrapper is `chpl_here_good_alloc_size()`

File: modules/internal/DefaultRectangular.chpl [hunk 2]
```diff
@@ -644,6 +644,13 @@ module DefaultRectangular {
     pragma "local field"
     var shiftedData : _ddata(eltType);
     var noinit_data: bool = false;
+
+    // 'dataAllocDom' is used by the array-vector operations (e.g. push_back,
+    // pop_back, insert, remove) to allow growing or shrinking the data
+    // buffer in a doubling/halving style.  If it is used, it will be the
+    // actual size of the data buffer, while 'dom' represents the size of the
+    // user-level array.
+    var dataAllocDom: domain(1);
```
Comment (bradcray) [inline modules/internal/DefaultRectangular.chpl]: Do we have any good way to tighten this up in order to avoid adding a whole 1D domain field to every default rectangular array that we allocate, whether it's an array-as-vec case or not?  Seems a bit clumsy and heavyweight.  Would it be possible to simply use an integer or a range for this purpose?  Do we know statically whether or not this field is used for a given array such that, if we had working void types, we could optimize the field away completely in the non-array-as-vec cases?

Comment (daviditen) [inline modules/internal/DefaultRectangular.chpl]: I just changed dataAllocDom into a range and am running it through paratest now.  I need it to track both a high and low bound, so a pair of integers would work, but a range's methods make it nicer.

We don't currently know statically if an array will be used as a vector in this way.  Any non-strided 1D Rectangular array with a unique domain (checked at runtime) can be used like this.

File: modules/internal/ChapelArray.chpl [hunk 3]
```diff
@@ -2027,7 +2030,28 @@ module ChapelArray {
...
+        }
+      } else {
+        // shrink to match the d2 bound on the side indicated by direction
+        if direction > 0 {
+          return lo..d2.high;
+        } else {
+          return d2.low..hi;
+        }
+      }
+    }
```
Comment (vasslitvinov) [inline modules/internal/ChapelArray.chpl]: I think we should have a blank line right before and right after resizeAllocRange() to separate it from neighbors. I am not seeing these lines. Please add.

File: modules/internal/DefaultRectangular.chpl [hunk 4]
```diff
@@ -644,6 +644,13 @@ module DefaultRectangular {
     pragma "local field"
     var shiftedData : _ddata(eltType);
     var noinit_data: bool = false;
+
+    // 'dataAllocRange' is used by the array-vector operations (e.g. push_back,
+    // pop_back, insert, remove) to allow growing or shrinking the data
+    // buffer in a doubling/halving style.  If it is used, it will be the
+    // actual size of the data buffer, while 'dom' represents the size of the
```
Comment (vasslitvinov) [inline modules/internal/DefaultRectangular.chpl]: I suggest putting "data" in "the data buffer" in single quotes, to suggest that that's the name of the field.

File: modules/internal/ChapelArray.chpl [hunk 5]
```diff
@@ -2027,7 +2030,28 @@ module ChapelArray {
     proc tail(): this._value.eltType {
       return this[this.domain.high];
     }
-
+    pragma "no doc"
+    proc resizeAllocRange(r: range, d2: domain, factor=arrayAsVecGrowthFactor, direction=1, grow=1) {
```
Comment (vasslitvinov) [inline modules/internal/ChapelArray.chpl]: Maybe make 'direction' and 'grow' param arguments and mark this function 'inline' ?

OTOH there is so much more going on in the callers of this function that the above may easily not make a difference.

Comment (vasslitvinov) [inline modules/internal/ChapelArray.chpl]: If we are making changes, I would add "compilerAssert(d2.rank == 1)" for documentation,
with a comment that "we are operating only on 1-D arrays"

File: modules/internal/ChapelArray.chpl [hunk 6]
```diff
@@ -2027,7 +2030,28 @@ module ChapelArray {
...
+      var newSize: int;
+      if grow > 0 {
+        newSize = max(size+1, (size*factor):int); // Always grow by at least 1.
+        if direction > 0 {
+          return lo..#newSize;
+        } else {
+          return ..hi#-newSize;
+        }
+      } else {
+        // shrink to match the d2 bound on the side indicated by direction
```
Comment (vasslitvinov) [inline modules/internal/ChapelArray.chpl]: Hm, does this has the effect that shrinking will typically be 1-by-1?

I'd suggest that if we are growing by 1.5, then we should shrink similarly.
Or maybe by a different factor, like 2 ?

File: modules/internal/ChapelArray.chpl [hunk 7]
```diff
@@ -2027,7 +2030,28 @@ module ChapelArray {
     proc tail(): this._value.eltType {
       return this[this.domain.high];
     }
-
```
Comment (vasslitvinov) [inline modules/internal/ChapelArray.chpl]: I'd like to see a comment saying something like:

Return a range that's a grow or a shrink of 'r' to accommodate 'd2'.

Comment (vasslitvinov) [inline modules/internal/ChapelArray.chpl]: There is a comment earlier in this file that needs updating:

       The current implementation reallocates the array every time the domain
       is modified.  This could be improved with a size doubling/halving
       strategy.

File: modules/internal/ChapelArray.chpl [hunk 8]
```diff
@@ -2027,7 +2030,28 @@ module ChapelArray {
...
+        if direction > 0 {
+          return lo..d2.high;
+        } else {
+          return d2.low..hi;
+        }
+      }
+    }
     /* Add element ``val`` to the back of the array, extending the array's
        domain by one. If the domain was ``{1..5}`` it will become ``{1..6}``.
      */
```
Comment (vasslitvinov) [inline modules/internal/ChapelArray.chpl]: While there, I suggest extending this comment (and similar comments on later functions) with:

The array must be a rectangular 1-D array; its domain must be non-stridable and not shared with other arrays."

File: modules/internal/ChapelArray.chpl [hunk 9]
```diff
@@ -2037,7 +2061,18 @@ module ChapelArray {
             hi = this.domain.high+1;
       const newDom = {lo..hi};
       on this._value {
-        this._value.dsiReallocate(newDom);
```
Comment (vasslitvinov) [inline modules/internal/ChapelArray.chpl]: While there - I would prefer 'newDom' to be a range. That will make the generate code slimmer+faster.
I do not see the need for 'newDom' to be a domain.

This could be done as a separate PR.

File: modules/internal/ChapelArray.chpl (hunk 6 above)
Comment (vasslitvinov) [inline modules/internal/ChapelArray.chpl]: Oh NM shrinking by a factor already happens - in the caller.

File: modules/internal/ChapelArray.chpl [hunk 10]
```diff
@@ -2037,7 +2061,18 @@ module ChapelArray {
...
+        if !this._value.dataAllocRange.member(hi) {
+          /* The new index is not in the allocated space.  We'll need to
+             realloc it. */
+          if this._value.dataAllocRange.length < this.domain.numIndices {
```
Comment (vasslitvinov) [inline modules/internal/ChapelArray.chpl]: Would this check be better as an assert, turned of upon --fast ?

Comment (vasslitvinov) [inline modules/internal/ChapelArray.chpl]: One way or another, it ought to be factored into a separate function, to improve readability of the code.

File: modules/internal/ChapelArray.chpl [hunk 11]
```diff
@@ -2053,7 +2088,13 @@ module ChapelArray {
             hi = this.domain.high-1;
       const newDom = {lo..hi};
       on this._value {
-        this._value.dsiReallocate(newDom);
```
Comment (vasslitvinov) [inline modules/internal/ChapelArray.chpl]: Please apply my comments above to all functions that they apply to.

File: modules/internal/DefaultRectangular.chpl [hunk 12]
```diff
@@ -644,6 +644,13 @@ module DefaultRectangular {
...
+    // user-level array.
+    var dataAllocRange: range(int);
```
Comment (vasslitvinov) [inline modules/internal/DefaultRectangular.chpl]: How about make the type of this field 'bool' (ideally 'void') if rank > 1 ? Save some memory.

File: modules/internal/ChapelArray.chpl (hunk 10 above)
Comment (daviditen) [inline modules/internal/ChapelArray.chpl]: If the array was being used as a non-vector-style array, the dataAllocRange can be out of date and smaller than this.domain.  In that case, it needs to be reset here.

File: modules/internal/DefaultRectangular.chpl (hunk 12 above)
Comment (daviditen) [inline modules/internal/DefaultRectangular.chpl]: I'd like to leave this for now.  Making it 'bool' would be misrepresenting what the field is for.  Once there is a way to make it 'void' I'll be OK with that.

File: modules/internal/ChapelArray.chpl (hunk 9 above)
Comment (daviditen) [inline modules/internal/ChapelArray.chpl]: 'dsiReallocate' takes a domain as an argument, so either 'newDom' needs to be a domain, or one needs to be created from it.

**Comment by daviditen** [comment]
I addressed most of the comments that I didn't respond to here in commit 26d9fdd.  Thanks for reviewing this, @vasslitvinov

File: modules/internal/ChapelArray.chpl (hunk 9 above)
Comment (daviditen) [inline modules/internal/ChapelArray.chpl]: I take back my previous comment about 'newDom'.  I see what you mean now and can clean that up in another PR.
